subiquity/server/controllers/ad.py

```diff
@@ -37,12 +38,45 @@
 FIELD = "Domain Controller name"
 
 
+class DcPingStrategy:
+    # Consider staging in the snap for future proof.
+    cmd = "/usr/sbin/realm"
+    arg = "discover"
+
+    async def ping(self, address: str) -> AdDomainNameValidation:
+        cp = await arun_command([self.cmd, self.arg, address], env={})
+        if cp.returncode:
+            return AdDomainNameValidation.REALM_NOT_FOUND
```

Comment:
Is the return value granular, meaning there may be other failure types, or is that pretty much the only result?

Reply:
A well formed `realm discover` has two possible exit codes: 0 for success and 1 for any kind of failure (from lack of network to DC not found):

https://gitlab.freedesktop.org/realmd/realmd/-/blob/master/tools/realm-discover.c#L154-163

So other failure scenarios would mean a bug in the formation of the command line or an exception raised by `arun_command` if the binary is not found.

Would you say we should consider those in the enum?

Reply:
OK, so if we're only assuming exit codes 0 or 1, this is fine.

> binary is not found

If we plan to sure that realm is present on the system, either by seeded as part of the installer squashes or in the snap, then that binary not found error shouldn't happen and can be an exception.  Same for bug in the formation of the command line.

This all looks good to me.

Reply:
If realm or a required binary is not available, then we shouldn't show the option to enable AD in the installer.

Reply:
That could be done, if desired.   https://github.com/canonical/subiquity/blob/main/DESIGN.md?plain=1#L242

Reply:
That means the client should `GET /active_directory` from this page to decide whether the `Use Active Directory` flag should be enabled or not, isn't it? I was considering implementing an endpoint for that, but raising is even easier, if that's the idiomatic approach, I'm ok with doing so.

![image](https://user-images.githubusercontent.com/11138291/219069437-b3efedfa-ce60-4809-adef-b0e25a369301.png)

Reply:
(As a side note I only see `raise Skip` in the client controllers, but not on the server ones).

Reply:
We did something like this recently, I'll get back to you.

Reply:
here's the bit that I was thinking of - commit fe4b6fde6fe6eb29db172c4fe9cb013d55ab5da5 - though that's client initiated so maybe not so similar.

If you want an endpoint the client can call to know if the screen can be shown then I think that's fine.

Reply:
I don't have strong opinion about this. @jpnurmi what do you think? We are discussing how the client would know ahead of time that the live system has the dependencies to configure AD, so it could disable the `Use Active Directory` check box accordingly. Options are:

1. `GET /active_directory` (before showing the active directory page) raises `Skip` if lacking support;
2. `GET /active_directory/has_support` (one extra endpoint - better naming suggestion is accepted) returns false if lacking support.

Reply:
I'll keep that in mind for my next PR that will address those small UX enhancements:

1. new endpoint to check ahead of time if AD is supported
2. auto discover the domain in the controller start